tests/nn/data_parallel/test_fsdp_uneven.py

```diff
@@ -29,27 +29,36 @@ def _test_func(rank, world_size, model, fsdp_config, tempfile_name, unused, test
 
     my_lr = 0.1
 
+    device = torch.device("cuda")
+    if fsdp_config.get("mixed_precision", False):
+        dtype = torch.float16
+        fsdp_config["fp32_reduce_scatter"] = True
```

Comment:
why is this needed?

Reply:
This change and the floats() below are related to mixed precision.

FSDP with `mixed_precision=True` will by default sum the grads in FP16 (reduce-scatter), then cast to FP32 and update the FP32 model params.

If I implement this in the “ref” case, then one of the tests fails in CI for one world size due to a difference of 2e-5. This is because a lot of the underlying math is happening in FP16 (e.g., grad sum), but because the end result is FP32, the default atol/rtol in assert_allclose are for FP32 which is too sensitive.

We could either adjust atol/rtol, or what I did instead is switch to a more precise mixed precision recipe: We cast the model output to FP32 to compute the loss (in this case just a sum). We then tell FSDP to do the reduce-scatter in FP32 (to match the ref case). Now all the tests pass with the original atol/rtol.